torch/fx/experimental/proxy_tensor.py

```diff
@@ -527,3 +528,55 @@ def wrap_fake_symbolic(x, sym_shape):
         return t
 
     return wrapped
+
+
+def get_torch_dispatch_modes():
+    modes = [torch._C._get_torch_dispatch_mode()]
+    if modes[-1] is None:
+        return list()
+    while modes[-1].inner is not None:
+        modes.append(modes[-1].inner)
+    return modes
+
+
+def get_isolated_graphmodule(func, args, kwargs):
+    """A helper function used to get the GraphModule for the given func.
+
+    It's expected to be used in the ProxyTensor tracing context.
+    It detaches the args and kwargs from the current tracer so that the trace of
+    the current graph module can be created without any side-effects.
+    """
+    # make_fx doesn't support kwargs, so we need to do this flattening
+    # and then unflatten the args before calling func
+    all_args, spec = pytree.tree_flatten((args, kwargs))
+
+    def wrapped(args):
+        fn_args, fn_kwargs = pytree.tree_unflatten(args, spec)
+        return func(*fn_args, **fn_kwargs)
+
+    unwrapped_all_args = [unwrap_elem(a) for a in all_args]
+
+    # Current implementation doesn't support the case when ProxyTensor is
+    # wrapped with another Tensor subclass
+    # See: https://github.com/pytorch/pytorch/pull/81764#issuecomment-1200472068
+    assert all(
+        getattr(a, "elem", None) is None
+        for a in unwrapped_all_args
+        if isinstance(a, torch.Tensor)
+    ), "ProxyTensor is wrapped with another Tensor subclass"
+
+
+    with contextlib.ExitStack() as stack:
+        modes = get_torch_dispatch_modes()
+        # Disable all torch dispatch modes
+        for mode in modes:
+            stack.enter_context(enable_torch_dispatch_mode(mode.inner, replace=mode))
+        assert torch._C._get_torch_dispatch_mode() is None
+
+        # Enable all torch dispatch modes except ProxyTorchDispatchMode
+        for mode in reversed([m for m in modes if not isinstance(m, ProxyTorchDispatchMode)]):
+            # mode.restore() doesn't work because mode.inner might be ProxyTorchDispatchMode
+            # mode.push() is restricted to modes that don't take any arguments
+            stack.enter_context(mode.push())
```

Comment:
This still seems super error prone. As you say here, this only works for modes that don't take any arguments. It would be much better if there was just a boolean toggle on the proxy dispatch mode you can use to turn it off without actually removing it from the stack.

Reply:
I just pushed a change to use `mode.restore()` with modified `mode.inner` and `mode.ancestors`, is this too hacky?

Reply:
I'm against anything that modifies mode.inner, we don't really know how to understand dynamically changing mode stack structure, it's really weird and hard to think about.

Reply:
Instead of the current approach, you would like to see the same as done with `in_kernel_invocation_manager` and `FakeTensorMode.in_kernel_invocation = True/False`? But it also mutates the attribute locally, yes not the `.inner` but a different attribute, still similar.

Reply:
Yes, that is my preference. The mutation in this case can be treated like a dynamically scoped variable with limited impact (as opposed to changing of inner, which totally changes the semantics of subsequent calls in the stack.)

Reply:
Okay, do you think changing of inner should be disallowed programmatically?

Reply:
In this situation, the semantics of modes is quite clear and asserted that outside the "with" block the modes are not changed. Isn't it a good thing to exercise various usage of `.inner` attribute? 
Is there an issue for 
https://github.com/pytorch/pytorch/blob/afafd16671b2d2f253cad7c3d453ad1fc44cccf5/torch/utils/_python_dispatch.py#L24-L25